File: CPP/Trie.cpp
#include<bits/stdc++.h>
using namespace std;

/*
    字典树(前缀数) Trie

    快速检索字符串
*/

class Node {
public:
    Node* children[26]{};
    bool end=false;
};

class Trie {
private:
    Node* root;
public:
    Trie() {
        root = new Node();
    }
    
    // 插入字符串到字典树中
    void insert(string word) {
        Node* node = root;
        int n = word.size();
        for( char& x : word ){
            int k = x - 'a';
            if( ! node -> children[k] ) node -> children[k] = new Node();
            node = node -> children[k];
        }
        node->end=true;
    }
    
    // 查找字典树中是否存在改字符串
    bool search(string word) {
        Node* node = root;
        int n = word.size();
        int i = 0;
        for( ; i < n ; i++ ) {
            int k = word[i] - 'a';
            if( ! node -> children[k] ) break;
            node = node -> children[k];
        }
        return i==n&&node->end;
    }
    
    // 查找是否存在该前缀
    bool startsWith(string prefix) {
        Node* node = root;
        int n = prefix.size();
        int i = 0;
        for( ; i < n ; i++ ) {
            int k = prefix[i] - 'a';
            if( ! node -> children[k] ) break;
            node = node -> children[k];
        }
        return i==n;
    }
};

File: Python/Trie.py

#
#   字典树(前缀数) Trie
#
#   快速检索字符串
#

class Node:
    def __init__(self):
        self.children = [None] * 26
        self.end = False

class Trie:

    def __init__(self):
        self.root = Node()

    # 插入字符串
    def insert(self, word: str) -> None:
        node = self.root
        for x in word:
            k = ord(x) - ord('a')
            if node.children[k] == None:
                node.children = Node()
            node = node.children[k]
        node.end = True

    # 搜索 字典树 中是否有完整的 word
    def search(self, word: str) -> bool:
        node = self.root
        i = 0
        for x in word:
            k = ord(x) - ord('a')
            if node.children[k] == None:
                break
            node = node.children[k]
            i+=1
        return i==n and node.end

    # 搜索 字典树 中是否存在 该前缀字符串
    def startsWith(self, prefix: str) -> bool:
        node = self.root
        i = 0
        for x in word:
            k = ord(x) - ord('a')
            if node.children[k] == None:
                break
            node = node.children[k]
            i+=1
        return i==n